[10.x] Add a `Number` utility class

emmadesilva opened the pull request:
## Abstract

Adds a new `Number` utility class, in the `Illuminate\Support` namespace, as suggested by Taylor in https://github.com/laravel/framework/pull/48827#issuecomment-1781232805.

```php
use Illuminate\Support\Number;

Number::format(123456789) // 123,456,789
Number::toCurrency(59.99) // $59.99
Number::bytesToHuman(1536) // 1.5 KB
Number::toHuman(10000) // ten thousand
```

### PR Development

The class ports the `bytesToHuman()` helper from https://github.com/laravel/framework/pull/48827, and after discussion here and in `#internals` adds some wrappers for the [`NumberFormatter`](https://www.php.net/manual/en/class.numberformatter.php) class. As this requires the `php-intl` extension, I added a check for those wrappers that throws a `RuntimeException`. I also added the extension as a suggestion to the `Support` package `composer.json`.

Please let me know if this route works, otherwise I'm happy to make changes.

**If anyone has ideas on more helpers, please let me know!**

### Future development

If this gets merged, I'll also make a PR that adds a `Numberable` (or something better named), similar to `Stringable`, that allows `Number::of(100)->toHuman()` and other neat stuff like that.

emmadesilva committed:
Create a new Number utility class

emmadesilva committed:
Add a `Number::bytesToHuman()` helper

Ports https://github.com/laravel/framework/pull/48827 into the new `Number` utility class

caendesilva commented on src/Illuminate/Support/Number.php:
```diff
@@ -0,0 +1,28 @@
+
+class Number
+{
+    use Macroable;
```
I added the `Macroable` trait, as the `Str` class uses it. Let me know if we should keep it.

henzeb replied:
We should. The Number helper has been requested many times before... This way people can add their functionality through packages before it's added to the core.

caendesilva replied:
Okay awesome! Resolving this conversation. Thanks :)

caendesilva commented:
I'm marking this ready for review, as I personally have no more ideas on what utilities to add. But if anyone has ideas I am more than happy to implement!

henzeb commented:
> I'm marking this ready for review, as I personally have no more ideas on what utilities to add. But if anyone has ideas I am more than happy to implement!

https://book.cakephp.org/2/en/core-libraries/helpers/number.html

realodix commented:
as suggested by Taylor in https://github.com/laravel/framework/pull/48827#issuecomment-1781232805

number_to_currency, number_to_human, number_to_human_size, number_to_percentage, number_with_delimiter, number_with_precision

caendesilva commented:
> > I'm marking this ready for review, as I personally have no more ideas on what utilities to add. But if anyone has ideas I am more than happy to implement!
> 
> as suggested by Taylor in [#48827 (comment)](https://github.com/laravel/framework/pull/48827#issuecomment-1781232805)
> 
> number_to_currency, number_to_human, number_to_human_size, number_to_percentage, number_with_delimiter, number_with_precision

Gonna take a stab at a `Number::toHuman()` helper now!

emmadesilva committed:
Add a `Number::toHuman()` helper

This is unfortunately dependent on the `ext-intl`, as it wraps the NumberFormatter class.

caendesilva commented on src/Illuminate/Support/Number.php:
```diff
@@ -2,12 +2,27 @@
 {
     use Macroable;
 
+    /**
+     * Format the number to a fluent human-readable string.
+     *
+     * @param float|int $number
+     * @param string $locale
+     * @return false|string
+     */
+    public static function toHuman($number, $locale = 'en')
+    {
+        $formatter = new NumberFormatter($locale, NumberFormatter::SPELLOUT);
+
+        return $formatter->format($number);
+    }
+
```
This is unfortunately dependent on the `ext-intl`, as it wraps the [NumberFormatter](https://www.php.net/manual/en/class.numberformatter.php) class. Assuming this is acceptable, we may want to add a check for the extension in the `NumberFormatter` class, and throw an exception if it is not available. But I understand if this is not worth it. However, given how much code a DIY solution would require, I personally think it is worth it.

caendesilva replied:
If this is acceptable, I'll go right ahead and add some more wrappers as the `NumberFormatter` class has a lot of useful helpers.

caendesilva replied:
@taylorotwell Hey! Hope it's okay to ping you. Just wanted your opinion regarding wrapping the `NumberFormatter` class, and how we should handle that.

marcovo commented:
The use of prefixes in `bytesToHuman()` seems to follow a custom naming scheme instead of established standards.

I notice two things:

1. The prefixes used seem to be [metrix prefixes](https://en.wikipedia.org/wiki/Metric_prefix) (`k`, `M`, `G`, ...), calculating base-10, with the exception that an uppercase `K` is used instead of the official lowercase `k`
2. The calculations seem to be done using base-2, which would actually suggest [binary prefixes](https://en.wikipedia.org/wiki/Binary_prefix) (`Ki`, `Mi`, `Gi`, ...) should be used

I propose to replace `bytesToHuman()` with two other methods:

1. `bytesToMetric()`, using division by 1000 and using metric abbreviations `kB`, `MB`, `GB`, ...
2. `bytesToBinary()` (or some better name), using division by 1024 and using binary abbreviations `KiB`, `MiB`, `GiB`, ...

caendesilva commented:
> The use of prefixes in `bytesToHuman()` seems to follow a custom naming scheme instead of established standards.
> 
> I notice two things:
> 
> 1. The prefixes used seem to be [metrix prefixes](https://en.wikipedia.org/wiki/Metric_prefix) (`k`, `M`, `G`, ...), calculating base-10, with the exception that an uppercase `K` is used instead of the official lowercase `k`
> 2. The calculations seem to be done using base-2, which would actually suggest [binary prefixes](https://en.wikipedia.org/wiki/Binary_prefix) (`Ki`, `Mi`, `Gi`, ...) should be used
> 
> I propose to replace `bytesToHuman()` with two other methods:
> 
> 1. `bytesToMetric()`, using division by 1000 and using metric abbreviations `kB`, `MB`, `GB`, ...
> 2. `bytesToBinary()` (or some better name), using division by 1024 and using binary abbreviations `KiB`, `MiB`, `GiB`, ...

Thank you for the feedback! You are correct that the `kB` should be lowercase. I will fix that. As for the rest, you are also correct that I am technically using binary abbreviations, as that's what computers use. I realise this makes the method quite opinionated, so let's await some extra feedback on that matter and see what others think.

emmadesilva committed:
Use lowercase `k` for kilobytes

See https://github.com/laravel/framework/issues/48845#issuecomment-1784984816

marcovo commented:
I want to stress that the code in the current state uses binary calculations in combination with decimal/metric abbreviations, a combination that would lead to technically confusing results. Do note however, that Windows (up to this date) also does exactly that: it reports filesizes in multiples of 1024 while using metric abbreviations. I would propose to not follow this false windows-path and stick to binary calculations with binary prefixes and/or metric calculations with metric prefixes.

emmadesilva committed:
Update Support package to suggest `ext-intl`

emmadesilva committed:
Throw if extension is not installed when using NumberFormatter wrapper

As discussed in #internals, this seems to be a good compromise when the extension is not used. Instead of testing against the exception in the tests, I just skipped the test if the extension is missing, as that is what we do in the InteractsWithRedis testing trait.

caendesilva commented:
> I want to stress that the code in the current state uses binary calculations in combination with decimal/metric abbreviations, a combination that would lead to technically confusing results. Do note however, that Windows (up to this date) also does exactly that: it reports filesizes in multiples of 1024 while using metric abbreviations. I would propose to not follow this false windows-path and stick to binary calculations with binary prefixes and/or metric calculations with metric prefixes.

Noted. I'm considering if we should add an option for if to use base 10 or base 2 but I'm weary of adding a bunch of complexity. Ideally I would like to get input from the core team on what they think is best.

emmadesilva committed:
Add a `Number::toCurrency()` helper

caendesilva commented on src/Illuminate/Support/Number.php:
```diff
@@ -32,7 +32,7 @@ public static function toHuman($number, $locale = 'en')
     public static function bytesToHuman($bytes, $precision = 2)
     {
-        $units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
+        $units = ['B', 'kB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
```
It's worth noting that Rails uses `KB`. https://api.rubyonrails.org/classes/ActionView/Helpers/NumberHelper.html

khalyomede commented:
Hi, congratulations on this PR, looks very useful I already can see some home made helpers I could replace by this one.

One question comes to my mind, will the toHuman() method will take into account the locale? I mean this:

```php
$number = Number::of(10);

$number->toCurrency("EUR"); // "€10.00"

App::setLocale("fr");

$number->toCurrency("EUR"); // "10,00 €" 
```

caendesilva commented:
> Hi, congratulations on this PR, looks very useful I already can see some home made helpers I could replace by this one.
> 
> One question comes to my mind, will the toHuman() method will take into account the locale? I mean this:
> 
> ```
> $number = Number::of(10);
> 
> $number->toCurrency("EUR"); // "€10.00"
> 
> App::setLocale("fr");
> 
> $number->toCurrency("EUR"); // "10,00 €" 
> ```

Glad you like it! Yes, someone on Twitter pointed this out, so I am just about to implement this.

emmadesilva committed:
Make Number helper locale parameters null and default to App locale

This makes so that if no locale parameter is specified, the configured App locale is used.

tresorkalanda-chd commented:
I'd also like it to have a helper that allows you to do this without importing the `Number` class directly.
```php
number()->toHuman();
```

emmadesilva committed:
Add a `Number::format()` helper

Adds a locale-aware number formatting helper

emmadesilva committed:
Fix number tests

Could not get Mockery to work, so this is my fix.

emmadesilva committed:
Add a `Number::toPercent()` helper

I'm dividing the supplied value by 100 so that 50 = 50% as that's how Rails does it, and that's what Taylor linked to in his suggestion for this class. The default number formatter would consider 0.5 to be 50% and 50 to be 5000%. I'm not sure which option is best, so I went with the Rails format. https://api.rubyonrails.org/classes/ActionView/Helpers/NumberHelper.html#method-i-number_to_percentage

caendesilva commented on src/Illuminate/Support/Number.php:
```diff
@@ -0,0 +1,113 @@
+
+        $formatter->setAttribute(NumberFormatter::MAX_FRACTION_DIGITS, $precision);
+
+        return $formatter->format($number / 100);
```
I'm dividing the supplied value by 100 so that 50 = 50% as that's how Rails does it, and that's what Taylor linked to in his suggestion for this class. The default number formatter would consider 0.5 to be 50% and 50 to be 5000%. I'm not sure which option is best, so I went with the Rails format. https://api.rubyonrails.org/classes/ActionView/Helpers/NumberHelper.html#method-i-number_to_percentage

caendesilva commented:
Okay, I think all the core methods are implemented, so I'm opening this back up for review.

utsavsomaiya commented:
You can also add this one!!

`Number::ordinal(1);`

<img width="867" alt="image" src="https://github.com/laravel/framework/assets/96036522/0484413a-b95f-4290-80ba-b0dca4fc5818">

caendesilva commented:
> You can also add this one!!
> 
> Ref: #48128
> 
> `Number::ordinal(1);`
> 
> <img alt="image" width="867" src="https://user-images.githubusercontent.com/96036522/280540805-0484413a-b95f-4290-80ba-b0dca4fc5818.png">

Good call, I like it. I understand Taylor was hesitant about it, but it may make more sense in this utility rather than the `Collection` class. Otherwise we can always remove it.

taylorotwell commented:
I think the `human` implementation is a bit better in Rails. This one where it totally spells out the word doesn't feel immediately useful.

emmadesilva committed:
Rename Number toHuman helper to spellout

We may want to remove it, as per https://github.com/laravel/framework/pull/48845#issuecomment-1802768912. But renaming it for now.

emmadesilva committed:
Create new `Number::toHuman()` helper

Based on the Rails implementation, as requested by Taylor in https://github.com/laravel/framework/pull/48845#issuecomment-1802768912

See https://api.rubyonrails.org/classes/ActionView/Helpers/NumberHelper.html#method-i-number_to_human

Uses the short scale system, see https://en.wikipedia.org/wiki/Long_and_short_scales

emmadesilva committed:
Change toHuman implementation to better match Rails version

Based more on the logic of Rails, but with added support for massive numbers.

emmadesilva committed:
Update toHuman helper to better handle extreme numbers

Inverts negative numbers, and removes unreachable cases, and handles very large numbers

emmadesilva committed:
Clean up toHuman helper

caendesilva commented:
> I think the `human` implementation is a bit better in Rails. This one where it totally spells out the word doesn't feel immediately useful.

Hey again Taylor! I changed the existing `human` implementation to be `spellout` just in case we want to keep it around. But I can totally remove it if there's no use for it.

I reimplemented the `toHuman` method to be based on the Rails logic, with some modifications to better handle extreme values. There is a lot of room for how to format extreme values both big and small, and I had initially some issues when handling integers over 64 bits, so I removed explicit support for quintillion values and up. 

```php
Number::toHuman(1000) // 1 thousand
Number::toHuman(1000000) // 1 million
Number::toHuman(1000000000) // 1 billion
```

Please let me know your thoughts on this, and if I should change anything!

taylorotwell committed:
formatting

taylorotwell committed:
formatting

taylorotwell committed:
formatting

taylorotwell committed:
formatting

taylorotwell committed:
formatting

taylorotwell commented:
Thanks 👍

OzanKurt commented on src/Illuminate/Support/Number.php:
```diff
@@ -0,0 +1,161 @@
+     * @param  int  $precision
+     * @return string
+     */
+    public static function forHumans(int|float $number, int $precision = 0)
```
What about other languages?

caendesilva replied:
Hmm, maybe the `__()` function could be used to provide translation support?

rmunate replied:
https://github.com/laravel/framework/pull/48845#issuecomment-1817535149

rmunate commented:
@caendesilva @taylorotwell  Hello Engineers, I would like to take into account the following comment, I would like to add a functionality that is called `Number::spell()` and that this returns an instance of my package **"https://github.com/rmunate/SpellNumber"** to translate to letters, to currency format and to ordinals.

I would like to be able to contemplate it, it would be something like this

```php
Number::spell(100)->locale('en')->toLetters();
// "One Hundred"
```

what do you think? Do you want me to work on adjusting this?

rmunate commented:
> I think the `human` implementation is a bit better in Rails. This one where it totally spells out the word doesn't feel immediately useful.

https://github.com/laravel/framework/pull/48845#issuecomment-1817535149

caendesilva commented:
> @caendesilva @taylorotwell Hello Engineers, I would like to take into account the following comment, I would like to add a functionality that is called `Number::spell()` and that this returns an instance of my package **"https://github.com/rmunate/SpellNumber"** to translate to letters, to currency format and to ordinals.

I'm not a maintainer so take this with a grain of salt, but I think what you are proposing would add too much complexity to be in the core framework, as most people don't need advanced spellout helpers, and those that do can use your package. Remember that this Number utility class is macroable, so your package could be updated to extend this class with your methods.

andrey-helldar commented:
It's fine for basic use, but if you need more flexible localization support, it's better to use the [`kwn/number-to-words`](https://github.com/kwn/number-to-words) package.

TursunboyevJahongir commented:
[this function](https://github.com/laravel/framework/pull/48845/commits/4b69ee5e7d3b4bf4ef8818861915cd38b5510262) works great, but can we use it with locate ?

```
Number::toHuman(1000, 'en') //1 thousand
Number::toHuman(1000, 'fr') //1 mille
```

jtibbles commented:
Hi, great work. However, regarding German formatting:
      Number::toCurrency(5.49, currency: 'EUR', locale: 'de') // 5.49

....in Germany (de_de) and Austria (de_at) they use a comma, not a dot. €5.49 is actually displayed as "€5,49".
Naturally it's always stored in a DB table with a dot (unless it's stored as a string) and converted to a dot for calculations.
Is there any chance this can be amended in a future update?

caendesilva commented:
> [this function](https://github.com/laravel/framework/pull/48845/commits/4b69ee5e7d3b4bf4ef8818861915cd38b5510262) works great, but can we use it with locate ?
> 
> ```
> Number::toHuman(1000, 'en') //1 thousand
> Number::toHuman(1000, 'fr') //1 mille
> ```

I like the idea and am gonna play around with it when I have some time over!

caendesilva commented:
> Hi, great work. However, regarding German formatting: Number::toCurrency(5.49, currency: 'EUR', locale: 'de') // 5.49
> 
> ....in Germany (de_de) and Austria (de_at) they use a comma, not a dot. €5.49 is actually displayed as "€5,49". Naturally it's always stored in a DB table with a dot (unless it's stored as a string) and converted to a dot for calculations. Is there any chance this can be amended in a future update?

Are you able to specify de_de/de_at as the locale? If those are incorrect, then this would be an bug in the upstream PHP extension.

jtibbles commented:
You can specify those and others that use the same format (cs_CZ for example). It's possible in the php.ini (setlocale(LC_MONETARY, 'cs_CZ') ). But many of my projects now expect multiple language packs now so while the amount and currency symbol stays the same, the formatting needs to be swap on the fly. Changing the locale for this is a bit cumbersome. It toCurrency can handle it instead it takes some weight off.
There's always str_replace I guess

caendesilva commented:
> You can specify those and others that use the same format (cs_CZ for example). It's possible in the php.ini (setlocale(LC_MONETARY, 'cs_CZ') ). But many of my projects now expect multiple language packs now so while the amount and currency symbol stays the same, the formatting needs to be swapped out on the fly. Changing the locale for this is a bit cumbersome. If toCurrency can handle it instead it takes some weight off. There's always str_replace I guess

You should probably be able to add a service provider or middleware to automatically set the `Number::useLocale()` to the request user's locale.

andrey-helldar commented:
> ....in Germany (de_de) and Austria (de_at) they use a comma, not a dot. €5.49 is actually displayed as "€5,49".
> Naturally it's always stored in a DB table with a dot (unless it's stored as a string) and converted to a dot for calculations.
> Is there any chance this can be amended in a future update?

@jtibbles, This is unlikely because Laravel does not provide translations. In fact, the [`toCurrency`](https://github.com/laravel/framework/pull/48845/files#diff-2866de1046fe4e657a94ff101ff6b049818966c89117bc87a9d2017c63315214R63-R70) method under the hood works with the [`NumberFormatter`](https://www.php.net/manual/en/class.numberformatter.php) class from PHP, and it accesses the `intl` library.

Therefore, there is no way to fix this error in Laravel until you replace `intl` with your own code. BUT you can use the [`kwn/number-to-words`](https://github.com/kwn/number-to-words) package - it knows how to format.

WikiRik commented:
@taylorotwell it seems that a part of the functionality was lost in one of your formatting commits; https://github.com/laravel/framework/pull/48845/commits/6d3a6c2affeb71c83d578fb2a18699b9699a4f87
Before this commit it used App::getLocale() to retrieve the default locale of the app and now it only has a new method called useLocale. Can you elaborate on why this change was made? Before people open a PR that adds the functionality to use the globally set locale (like was expected by Povilas of Laravel Daily in this video; https://www.youtube.com/watch?v=PRhhqukK9uw )

caendesilva commented:
> @taylorotwell it seems that a part of the functionality was lost in one of your formatting commits; [6d3a6c2](https://github.com/laravel/framework/commit/6d3a6c2affeb71c83d578fb2a18699b9699a4f87) Before this commit it used App::getLocale() to retrieve the default locale of the app and now it only has a new method called useLocale. Can you elaborate on why this change was made? Before people open a PR that adds the functionality to use the globally set locale (like was expected by Povilas of Laravel Daily in this video; https://www.youtube.com/watch?v=PRhhqukK9uw )

Thanks for bringing this up! I've been wondering this too, especially since it's something I've been asked about a bunch.

andrersilva96 commented:
Hi @caendesilva I was testing with brazilian currency, I believe that we use comma to separate decimals and dot for thousands:
R$11,000.36  // \Illuminate\Support\Number::currency(11000.36, 'BRL')
R$ 11.000,36 // using my own helper

caendesilva commented:
> Hi @caendesilva I was testing with brazilian currency, I believe that we use comma to separate decimals and dot for thousands: R$11,000.36 // \Illuminate\Support\Number::currency(11000.36, 'BRL') R$ 11.000,36 // using my own helper

Hey, @andrersilva96 thanks for bringing it up! So I took a quick look, and does seem like the Brazilian reals are indeed divided using commas (inverted behaviour compared to the US dollar formatting). Though I did not in my brief search find an official source.

However, this helper is a wrapper for the PHP Intl extension, so this means that the bug is present in the PHP Core and needs to be fixed there. I suggest you file a bug report through here https://www.php.net/manual/en/install.problems.bugs.php. Feel free to ping me in the report!

becker commented:
@andrersilva96, I believe you could do this way:

```php
Number::currency(11000.36, 'BRL', 'pt_BR')
// "R$ 11.000,36"
```

andrersilva96 commented:
@becker This worked, thank's a lot ou melhor obrigado meu conterrâneo! haha

shaedrich commented:
From rmunate/SpellNumber#24 after [the repository has been discontinued](https://github.com/rmunate/SpellNumber/commit/22b4bd162b016628f6f8a58bccd53f289611e1f9):

![grafik](https://github.com/rmunate/SpellNumber/assets/11225821/9175780f-aa7b-4e0f-94aa-77726517609d)
(see [Appendix:English numerals](https://en.wiktionary.org/wiki/Appendix:English_numerals))

\+ [Scientific notation](https://en.wikipedia.org/wiki/Scientific_notation)/[engineering notation](https://en.wikipedia.org/wiki/Engineering_notation)/E notation

```php
$input = SpellNumber::integer(2);
$input->toLatinateOrdinal(); // "secondary"
$input->toAdverbial(); // "twice" as in "twice as good"
$input->toMultiplier(); // "twofold"
$input->toLatinateMultiplier(); // "double" as in "double the amount"
$input->toDistributive(); // "doubly"
$input->toMetricCollectivePrefix(); // "double-"
$input->toLatinateCollectivePrefix(); // "bi-" as in "bi-weekly"
$input->toFractional(); // "half" as in "half as good"
$input->toMetricFractionalPrefix(); // "demi-" as in "demi-god"
$input->toLatinateFractionalPrefix(); // "semi-" as in "semipermeable"
```

```php
$input = SpellNumber::value(3200);
$input->toScientificNotation(); // 3.2e3 or 3.2e-3 or 3.2E-3
$input->toEngineeringNotation(); // aliases
$input->toENotation(); // aliases
```

olivermbs commented:
I believe we lost the functionality to spell currencies which was available in [rmunate/SpellNumber](https://github.com/rmunate/SpellNumber), is that correct? I cannot seem to find a way achieve this with Illuminate\Support\Number

```php
SpellNumber::value(100)->locale('en')->currency('Dollars')->toMoney();
// "one hundred dollars"

rmunate commented:
> I believe we lost the functionality to spell currencies which was available in [rmunate/SpellNumber](https://github.com/rmunate/SpellNumber), is that correct? I cannot seem to find a way achieve this with Illuminate\Support\Number
> 
> ```
> SpellNumber::value(100)->locale('en')->currency('Dollars')->toMoney();
> // "one hundred dollars"
> ```

Hello, with Laravel's built-in class, you could try something like this:

```php
Number::spell(100, 'en') . " dollars";  
// "one hundred dollars"
```

ArnaudLier commented on src/Illuminate/Support/Number.php:
```diff
@@ -0,0 +1,161 @@
+     * @param  int  $precision
+     * @return string
+     */
+    public static function toFileSize(int|float $bytes, int $precision = 0)
+    {
+        $units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
+
+        for ($i = 0; ($bytes / 1024) > 0.9 && ($i < count($units) - 1); $i++) {
+            $bytes /= 1024;
+        }
+
+        return sprintf('%s %s', number_format($bytes, $precision), $units[$i]);
+    }
```
Was the idea to use the binary prefixes (KiB, MiB, etc.) abandoned?
